### PR TITLE
First attempt at removing DropAndReplace

### LINKED ISSUE #70919: spurious "borrow might be used .. when [variable] is dropped and runs the destructor"
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ca0e9f5171bf484d9b744d08057203ec)
```rust
use std::rc::Rc;
use std::cell::{RefCell, RefMut};
use std::mem::drop;

fn huh(x: Rc<RefCell<i32>>) {
    let mut rc: Rc<RefCell<i32>> = x.clone();
    let mut inner: RefMut<'_, i32> = rc.borrow_mut();
    loop {
        drop(inner);
        drop(rc);
        rc = x.clone();
        inner = rc.borrow_mut();
    }
}
```
produces:
```
error[E0505]: cannot move out of `rc` because it is borrowed
  --> src/lib.rs:9:14
   |
6  |     let mut inner: RefMut<'_, i32> = rc.borrow_mut();
   |                                      -- borrow of `rc` occurs here
...
9  |         drop(rc);
   |              ^^ move out of `rc` occurs here
...
12 |         inner = rc.borrow_mut();
   |         ----- borrow might be used here, when `inner` is dropped and runs the destructor for type `std::cell::RefMut<'_, i32>`

error[E0506]: cannot assign to `rc` because it is borrowed
  --> src/lib.rs:11:9
   |
6  |     let mut inner: RefMut<'_, i32> = rc.borrow_mut();
   |                                      -- borrow of `rc` occurs here
...
11 |         rc = x.clone();
   |         ^^ assignment to borrowed `rc` occurs here
12 |         inner = rc.borrow_mut();
   |         ----- borrow might be used here, when `inner` is dropped and runs the destructor for type `std::cell::RefMut<'_, i32>`
```

(Same error with `-Zpolonius`.)

The error messages are wrong: the assignment to `inner` can never run a destructor, since the previous value of `inner` was dropped earlier on.

And AFAICT this code is sound and ought to be accepted.

### PR DESCRIPTION
**Pull request by zeegomo**
First attempt at removing the `DropAndReplace` terminator which complicates things a bit in some cases.
Based on  [zulip/drop-to-drop-if](https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/.60DROP.60.20to.20.60DROP_IF.60.20compiler-team.23558)

Eventually, I re-introduced a flag in the `Drop` terminator to indicate whether this drop is from a drop and replace operation. This is used mainly for better diagnostic and to check some invariants.

## Failing UI Tests

* ### Diagnostic underline whole assignment instead of just the left value
ex: 
```bash
71	LL |     *ap0 = ap1;
-	   |     ^^^^ assignment requires that `'2` must outlive `'1`
+	   |     ^^^^^^^^^^ assignment requires that `'2` must outlive `'1`
73	   |
```
[ui] src/test/ui/borrowck/borrowck-field-sensitivity.rs
[ui] src/test/ui/borrowck/borrowck-issue-48962.rs
[ui] src/test/ui/borrowck/borrowck-loan-of-static-data-issue-27616.rs
[ui] src/test/ui/borrowck/borrowck-partial-reinit-1.rs
[ui] src/test/ui/borrowck/borrowck-partial-reinit-2.rs
[ui] src/test/ui/borrowck/borrowck-partial-reinit-4.rs
[ui] src/test/ui/borrowck/borrowck-vec-pattern-nesting.rs
[ui] src/test/ui/borrowck/index-mut-help.rs
[ui] src/test/ui/borrowck/issue-45199.rs
[ui] src/test/ui/c-variadic/variadic-ffi-4.rs
[ui] src/test/ui/closures/2229_closure_analysis/diagnostics/box.rs
[ui] src/test/ui/liveness/liveness-assign/liveness-assign-imm-local-with-drop.rs
[ui] src/test/ui/nll/issue-27868.rs
[ui] src/test/ui/object-lifetime/object-lifetime-default-from-box-error.rs
[ui] src/test/ui/regions/regions-infer-paramd-indirect.rs
[ui] src/test/ui/unboxed-closures/unboxed-closures-failed-recursive-fn-1.rs

* ### Replace kills loans on previous value
```bash
12	LL |     println!("thread result: {:?}", res);
13	   |                                     --- borrow later used here
14	
-	error[E0505]: cannot move out of `greeting` because it is borrowed
-	  --> $DIR/issue-58776-borrowck-scans-children.rs:7:10
-	   |
-	LL |     let res = (|| (|| &greeting)())();
-	   |                --      -------- borrow occurs due to use in closure
-	   |                |
-	   |                borrow of `greeting` occurs here
-	...
-	LL |     drop(greeting);
-	   |          ^^^^^^^^ move out of `greeting` occurs here
-	...
-	LL |     println!("thread result: {:?}", res);
-	   |                                     --- borrow later used here
+	error: aborting due to previous error
28	
-	error: aborting due to 2 previous errors
-	
-	Some errors have detailed explanations: E0505, E0506.
-	For more information about an error, try `rustc --explain E0505`.
+	For more information about this error, try `rustc --explain E0506`.
33	
```
destructured assignment from drop and replace now kills loans. I don't think it's too big of an issue as the assignment should already trigger the borrowck, but the compiler now reports fewer errors in some cases
[ui] src/test/ui/borrowck/issue-58776-borrowck-scans-children.rs

Also fixes https://github.com/rust-lang/rust/issues/70919 but I didn't add any regression tests for it, might borrow them from https://github.com/rust-lang/rust/pull/102078 in a separate PR once this lands

### CODE REVIEW

**Comment by Zeegomo** [comment]
Figured out what the problem with the future size was.
The desugaring of `a = b;` in `drop(a); assign(a, b)` at MIR creation might create dead MIR blocks on the unwind path, because the drop is a no-op (e.g. the variable is certainly dead at this stage) or can't unwind.
During drop elaboration, we run some dataflow analyses to track the initialization state of the variables to drop and here's the problem.
For some reason (would love some insight), we push all MIR blocks in the dataflow work queue and process them, even thought they might not be reachable from the start (just talking about forward analyses).
In this case, this cause processing of dead blocks which propagate incorrect values for liveness analysis.
My solution was to change the fixpoint alg to only consider blocks reachable from the start.
For backwards compatibility this can be tweaked with a flag, might be useful for other analyses as well.

File: compiler/rustc_borrowck/src/invalidation.rs [hunk 1]
```diff
@@ -109,22 +112,36 @@ impl<'cx, 'tcx> Visitor<'tcx> for InvalidationGenerator<'cx, 'tcx> {
...
+                        .get(*target)
+                        .expect("MIR should be complete at this point")
+                        .statements
+                        .first()
+                } else {
+                    None
+                };
+
+                match next_statement {
+                    Some(Statement { kind: StatementKind::Assign(_), source_info: _ }) => {
```
Comment (nikomatsakis) [inline compiler/rustc_borrowck/src/invalidation.rs:128]: I don't think this check is sufficient -- you are just checking that the DROP is followed by an ASSIGNMENT, but actually don't you want to check that the place being dropped or assigned is the same?

[This code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b98f6d962782c5c8a5b0389a342ebb2c) for example yields MIR like

```
bb1: {
        drop(_1) -> bb2;                 // scope 0 at [src/main.rs:4:5: 4:6](https://play.rust-lang.org/#)
    }

    bb2: {
        _2 = String::new() -> bb3;       // scope 0 at [src/main.rs:5:13: 5:26](https://play.rust-lang.org/#)
                                         // mir::Constant
                                         // + span: [src/main.rs:5:13: 5:24](https://play.rust-lang.org/#)
                                         // + literal: Const { ty: fn() -> String {String::new}, val: Value(<ZST>) }
    }
```

which I think would trigger here, but shouldn't...?

Oh, I guess it's gated by the "is-replace" flag on `Drop`, so that probably works ok. Hmm. It still seems a bit sketchy to me.

In general, I am wondering, why are the diagnostics better if we do a `mutate_place` call here? I would rather modify the way that `access_place` *reports* the error (but not the check itself). 

Also, if `is_replace` is true, shouldn't the next statement *always* be an assignment? i.e., we could replace this with an assertion.

File: compiler/rustc_borrowck/src/lib.rs [hunk 2]
```diff
@@ -647,29 +652,43 @@ impl<'cx, 'tcx> rustc_mir_dataflow::ResultsVisitor<'cx, 'tcx> for MirBorrowckCtx
...
-            }
-            TerminatorKind::DropAndReplace {
-                place: drop_place,
-                value: ref new_value,
-                target: _,
-                unwind: _,
-            } => {
-                self.mutate_place(loc, (drop_place, span), Deep, flow_state);
-                self.consume_operand(loc, (new_value, span), flow_state);
+                let next_statement = if is_replace {
```
Comment (nikomatsakis) [inline compiler/rustc_borrowck/src/lib.rs:662]: As above. 

(Also, why do we apparently have the same code twice? Huh. I don't remember that.)

File: compiler/rustc_borrowck/src/used_muts.rs [hunk 3]
```diff
@@ -71,9 +71,10 @@ impl<'visit, 'cx, 'tcx> Visitor<'tcx> for GatherUsedMutsVisitor<'visit, 'cx, 'tc
             TerminatorKind::Call { destination, .. } => {
                 self.remove_never_initialized_mut_locals(*destination);
             }
-            TerminatorKind::DropAndReplace { place, .. } => {
-                self.remove_never_initialized_mut_locals(*place);
-            }
+            // FIXME: ??
+            // TerminatorKind::DropAndReplace { place, .. } => {
+            //     self.remove_never_initialized_mut_locals(*place);
+            // }
```
Comment (nikomatsakis) [inline compiler/rustc_borrowck/src/used_muts.rs:77]: I think can just remove this comment. The same effect is achieved by the `visit_statement` call below, which sees the `Assign`

File: compiler/rustc_mir_dataflow/src/framework/engine.rs [hunk 4]
```diff
@@ -99,13 +107,18 @@ where
     T: Idx,
 {
     /// Creates a new `Engine` to solve a gen-kill dataflow problem.
-    pub fn new_gen_kill(tcx: TyCtxt<'tcx>, body: &'a mir::Body<'tcx>, analysis: A) -> Self {
+    pub fn new_gen_kill(
+        tcx: TyCtxt<'tcx>,
+        body: &'a mir::Body<'tcx>,
+        analysis: A,
+        blocks: Blocks,
```
Comment (nikomatsakis) [inline compiler/rustc_mir_dataflow/src/framework/engine.rs:114]: Why did we have to add this flag? I suspect this is related to the unwinding comments that I added above, though I'm not sure.

File: compiler/rustc_mir_transform/src/dest_prop.rs [hunk 5]
```diff
@@ -544,23 +546,24 @@ impl<'a> Conflicts<'a> {
...
+            //         && !place.is_indirect()
+            //         && !dropped_place.is_indirect()
+            //     {
+            //         self.record_local_conflict(
+            //             place.local,
+            //             dropped_place.local,
+            //             "DropAndReplace operand overlap",
+            //         );
+            //     }
+            // }
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply compiler/rustc_mir_transform/src/dest_prop.rs:566]: @oli-obk can perhaps confirm this too

File: compiler/rustc_mir_transform/src/elaborate_drops.rs [hunk 6]
```diff
@@ -379,112 +377,38 @@ impl<'b, 'tcx> ElaborateDropsCtxt<'b, 'tcx> {
...
+                                TerminatorKind::Drop {
+                                    place,
+                                    target,
+                                    unwind: if data.is_cleanup {
+                                        None
+                                    } else {
+                                        Some(Option::unwrap_or(unwind, resume_block))
+                                    },
+                                    is_replace,
+                                },
```
Comment (nikomatsakis) [inline compiler/rustc_mir_transform/src/elaborate_drops.rs:402]: I don't follow this logic

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Initial round of comments

File: compiler/rustc_mir_transform/src/elaborate_drops.rs (hunk 6 above)
Comment (Zeegomo) [reply compiler/rustc_mir_transform/src/elaborate_drops.rs:402]: It's this case here https://github.com/rust-lang/rust/blob/270c94e484e19764a2832ef918c95224eb3f17c7/compiler/rustc_mir_transform/src/elaborate_drops.rs#L477, will add a better comment

File: compiler/rustc_borrowck/src/invalidation.rs (hunk 1 above)
Comment (Zeegomo) [reply compiler/rustc_borrowck/src/invalidation.rs:128]: > Oh, I guess it's gated by the "is-replace" flag on Drop, so that probably works ok. Hmm. It still seems a bit sketchy to me.

Yep

> In general, I am wondering, why are the diagnostics better if we do a mutate_place call here? I would rather modify the way that access_place reports the error (but not the check itself).

I would be very happy to do that. I do not remember now exactly what were the differences, but it was also to try push this to be as close to what was done before as possible. I'll try to see exactly what the error messages would look like and if we can improve the `access_place` reporting.

> Also, if is_replace is true, shouldn't the next statement always be an assignment? i.e., we could replace this with an assertion.

Yes, I was not sure at first due to possible optimizations

File: compiler/rustc_mir_dataflow/src/framework/engine.rs [hunk 7]
```diff
@@ -197,21 +217,32 @@ where
...
+            match blocks {
+                Blocks::All => {
+                    for (bb, _) in traversal::reverse_postorder(body) {
+                        dirty_queue.insert(bb);
+                    }
+                }
+                Blocks::Reachable => {
+                    dirty_queue.insert(mir::START_BLOCK);
+                }
             }
```
Comment (tmiasko) [inline compiler/rustc_mir_dataflow/src/framework/engine.rs:238]: The existing implementation already considers only reachable blocks, since `reverse_postorder` and `postorder` return only reachable blocks. Furthermore, is an error to query data flow state for unreachable blocks (there is a debug only assertion).

Though, it looks like the initial seeding of the work queue uses a different definition of reachability that the one in fixpoint computation. In particular, the fixpoint iteration ignores some unwind edges as configured by `dead_unwinds` method. Maybe this is the cause of the issue you described in https://github.com/rust-lang/rust/pull/104488#issuecomment-1365802291?

File: compiler/rustc_mir_build/src/build/scope.rs [hunk 8]
```diff
@@ -1094,21 +1094,35 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
-            TerminatorKind::DropAndReplace { place, value, target: next_target, unwind: None },
+            TerminatorKind::Drop {
+                place,
+                target: next_target,
+                unwind: Some(assign),
+                is_replace: true,
+            },
         );
+
         self.diverge_from(block);
```
[the remark this answers is not in the document]
Comment (Zeegomo) [reply compiler/rustc_mir_build/src/build/scope.rs:1126]: > I think that diverge_from will place block in a list and fix the terminator of assign later. Kind of confusing.

Yes, I initially considered the suggestion in the last paragraph of your comment but wanted to avoid unnecessary allocations. I guess this option can be made clearer with an obviously dummy target for the goto and a comment.

>Also, I presume that the goal of the diverge_from mechanism is to avoid creating unwind blocks when we are in panic=abort mode.

Good catch, I need to check, I don't remember something explicit about the panic strategy during mir build though, maybe it's shadowed by some other logic

File: compiler/rustc_mir_dataflow/src/framework/engine.rs (hunk 4 above)
Comment (Zeegomo) [reply compiler/rustc_mir_dataflow/src/framework/engine.rs:114]: It's related to https://github.com/rust-lang/rust/pull/104488#issuecomment-1365802291. 
In the fixpoint alg we push all mir blocks in the work queue, even if they are not reachable from the start for a forward analysis, essentially negating the effects of the `dead_unwind` flag in some cases. This is undesirable for drop elaboration due to the desugaring of `DropAndReplace` so I added a flag to control whether we initially insert all blocks in the work queue. 
All other analyses have the same behavior as before but maybe some could benefit from this optimization as well.

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 9]
```diff
@@ -2200,10 +2200,10 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
                             StorageDeadOrDrop::LocalStorageDead
                             | StorageDeadOrDrop::BoxedStorageDead => {
-                                assert!(
-                                    place_ty.ty.is_box(),
-                                    "Drop of value behind a reference or raw pointer"
-                                );
+                                // assert!(
+                                //     place_ty.ty.is_box(),
+                                //     "Drop of value behind a reference or raw pointer"
+                                // );
```
[the remark this answers is not in the document]
Comment (Zeegomo) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:2206]: It was a reminder to check why I needed to change this. It's probably something like https://github.com/rust-lang/rust/blob/270c94e484e19764a2832ef918c95224eb3f17c7/compiler/rustc_mir_transform/src/elaborate_drops.rs#L477 but will check

File: compiler/rustc_mir_dataflow/src/framework/engine.rs (hunk 4 above)
Comment (nikomatsakis) [reply compiler/rustc_mir_dataflow/src/framework/engine.rs:114]: I'm confused -- why does the desugaring of drop-and-replace have an effect here? The only reason I can think of that it would generate a lot of dead code would be if you had panic=abortmode?

File: compiler/rustc_mir_build/src/build/scope.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply compiler/rustc_mir_build/src/build/scope.rs:1126]: Why would this create more allocations? I haven't read into the code, but I imagine there is a `Vec<DivergeFrom>` stored *somewhere*, isn't it just going to make `DivergeFrom` a bit larger (because it now stores a `Option<(Place, Rvalue)>` pair or something like that)

[the remark this answers is not in the document]
Comment (nikomatsakis) [reply compiler/rustc_mir_build/src/build/scope.rs:1126]: (That said, that might make the data structure significantly bigger.)

File: compiler/rustc_mir_dataflow/src/framework/engine.rs (hunk 7 above)
Comment (tmiasko) [reply compiler/rustc_mir_dataflow/src/framework/engine.rs:238]: If this indeed the issue, I would suggest removing dead unwind edges directly in MIR and getting rid of `Engine::dead_unwinds` functionality.

Comment (Zeegomo) [reply compiler/rustc_mir_dataflow/src/framework/engine.rs:238]: > Though, it looks like the initial seeding of the work queue uses a different definition of reachability that the one in fixpoint computation. In particular, the fixpoint iteration ignores some unwind edges as configured by dead_unwinds method

Yep, that's exactly what I meant

> The existing implementation already considers only reachable blocks, since reverse_postorder and postorder return only reachable blocks. Furthermore, is an error to query data flow state for unreachable blocks

Oh, interesting. I remember seeing an error somewhere when I made `Blocks::Reachable` the default but can't seem to replicate it now.

> If this indeed the issue, I would suggest removing dead unwind edges directly in MIR and getting rid of Engine::dead_unwinds functionality.

That would be nice but I'm trying to keep modifications to a minimum here to maintain the focus and to me it seems a bit more work. I think a reasonably clean and easy change would be indeed to just put `mir::START_BLOCK` inside the queue at startup, what do you think?

File: compiler/rustc_mir_dataflow/src/framework/engine.rs (hunk 4 above)
Comment (Zeegomo) [reply compiler/rustc_mir_dataflow/src/framework/engine.rs:114]: Consider something like 
```rust
fn single() {
    let x;
    noop();
    x = Big::new();
}
```
The `maybe_init` analisys on the MIR right before drop elaboration will look like (without any modifications to the dataflow engine): 


The assignment `x = Big::new()` is now desugared into `drop(_1)` in bb2 and  `_1 = move _3` in bb3 and bb4.
Now, that `drop(_1)` in bb2 is a noop since `_1` was not assigned before, and bb4 is unreachable, but drop elaboration hasn't run and removed it yet. 
`drop(_3)` in bb3 is also a noop because we just moved out of `_3`.
However, by pushing all blocks in the dataflow engine work queue, bb4 might propagate its effects to its successors, in this case the variable `_1` being live, which might pollute other legit unwind blocks like bb7.
What we end up with is incorrect init data in bb7, since at that point in the CFG we can't possible have `_1` live.

If instead we only push in the queue blocks that are actually reachable from the start bb4 will not be considered and will not influence other blocks.

Comment (Zeegomo) [reply compiler/rustc_mir_dataflow/src/framework/engine.rs]: Anyway, I think we can remove the flag and let this be the default behavior for forward analyses. I think it's already what the user mean by providing `dead_unwinds`

File: compiler/rustc_borrowck/src/invalidation.rs (hunk 1 above)
Comment (Zeegomo) [reply compiler/rustc_borrowck/src/invalidation.rs:128]: >In general, I am wondering, why are the diagnostics better if we do a mutate_place call here? I would rather modify the way that access_place reports the error (but not the check itself).

For the user I think it's easier to report a problem in the assignment (i.e. a mutate) rather than the implicit drop in the case of a replace. We could also report both but I think that will clutter the report unnecessarily.

Btw, is there a logical difference between `mutate_place` and `access_place` with write access? Looks like `mutate_place` is mostly a shorthand for `access_place` with a few implicit arguments. Changing the write kind from `StorageDeadOrDrop` to `Mutate` will make the error reporting in `access_place`  more user friendly for a replace and shouldn't change any logic if I understand the comment on `WriteKind` correctly.

File: compiler/rustc_mir_build/src/build/scope.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Zeegomo) [reply compiler/rustc_mir_build/src/build/scope.rs:1129]: > Also, I presume that the goal of the diverge_from mechanism is to avoid creating unwind blocks when we are in panic=abort mode.

after a quick investigation (MIR dump of the nightly compiler) it looks like unwind/cleanup blocks are always created and later removed by the `AbortUnwindingCalls` pass.

I changed the terminator to `TerminatorKind::Unreachable` and added a comment to make it clearer that it's only temporary

File: compiler/rustc_borrowck/src/invalidation.rs (hunk 1 above)
Comment (Zeegomo) [reply compiler/rustc_borrowck/src/invalidation.rs]: In the end, I've tried another option which is simply to skip the drop access check in case of a replace and rely on the assignment to report possible problems. One thing to handle separately is the check access depth, which in case of an assignment following a skipped drop access should be `AccessDepth::Drop`.

**Comment by Zeegomo** [comment]
Some of the changes here like https://github.com/rust-lang/rust/pull/104488/commits/114c696577e7bb205d04a9463b31d7a85664c930 could be moved to a separate independent PR. 
Not sure of what's your preference on this, didn't want to split the discussion in too many places

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 10]
```diff
@@ -578,44 +577,7 @@ pub enum TerminatorKind<'tcx> {
...
-    /// ```
-    ///
-    /// Disallowed after drop elaboration.
-    DropAndReplace {
-        place: Place<'tcx>,
-        value: Operand<'tcx>,
-        target: BasicBlock,
-        unwind: Option<BasicBlock>,
-    },
+    Drop { place: Place<'tcx>, target: BasicBlock, unwind: Option<BasicBlock>, is_replace: bool },
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/syntax.rs:580]: The comment above needs to explain what effect (if any) `is_replace` has on the semantics of this operation

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Zeegomo) [reply compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs]: removed the comment

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 11]
```diff
@@ -578,43 +577,13 @@ pub enum TerminatorKind<'tcx> {
...
-    ///
-    /// Disallowed after drop elaboration.
-    DropAndReplace {
+    Drop {
         place: Place<'tcx>,
-        value: Operand<'tcx>,
         target: BasicBlock,
         unwind: Option<BasicBlock>,
+        /// This field is only used for better diagnostic and to check MIR drop invariants.
+        /// It has no effect on the semantics of the `Drop` terminator.
```
Comment (Aaron1011) [inline compiler/rustc_middle/src/mir/syntax.rs:585]: Aren't Drop invariants are part of the semantics of the `Drop` terminator? If I understand correctly, incorrectly setting this flag to `true` could cause invalid mir to be accepted by the borrow checker (if we assume that there's a later assignment when there actually isn't).

Comment (Zeegomo) [reply compiler/rustc_middle/src/mir/syntax.rs:585]: Mmm, good question.
> incorrectly setting this flag to true could cause invalid mir to be accepted by the borrow checker (if we assume that there's a later assignment when there actually isn't).

I would say that, by definition, changing the value of `is_replace` shouldn't change any property of a program.
If it does, it's because of implementation details (which are not relevant at this level).
We could even remove the field and check whether the next statement is an assignment to the same place everywhere we need the value of the flag.

The behavior in the borrowck is a result of the current implementation, not intrinsic in the definition of the `Drop` terminator. (e.g. my previous implementation didn't have this problem and the definition of `Drop` hasn't changed). In other words, the borrowck should check the access regardless of `is_replace`.


Actually, I would say that the semantics of `Drop` being dependent on `is_replace` would suggest that the removal of `DropAndReplace` was not very effective

Comment (oli-obk) [reply compiler/rustc_middle/src/mir/syntax.rs:585]: To avoid such issues and keep good diagnostics at the same time, you could change mir building to add some information to the spans that are used for drop terminators: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html#method.mark_with_reason

File: compiler/rustc_mir_build/src/build/scope.rs [hunk 12]
```diff
@@ -1118,21 +1118,38 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+        value: Rvalue<'tcx>,
     ) -> BlockAnd<()> {
         let source_info = self.source_info(span);
         let next_target = self.cfg.start_new_block();
 
+        let assign = self.cfg.start_new_cleanup_block();
+        self.cfg.push_assign(assign, statement_source_info, place, value.clone());
+        // We still have to build the scope drops so we don't know which block will follow.
+        // This terminator will be overwritten once the unwind drop tree builder runs.
+        self.cfg.terminate(assign, statement_source_info, TerminatorKind::Unreachable);
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/scope.rs:1140]: Is this necessary? Since the terminator will be added later, it can be left "uninitialized" (it is internally an `Option::None`)

File: compiler/rustc_const_eval/src/transform/validate.rs [hunk 13]
```diff
@@ -792,18 +792,19 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
...
+            //         self.fail(
+            //             location,
+            //             "`DropAndReplace` should have been removed during drop elaboration",
+            //         );
+            //     }
+            //     self.check_edge(location, *target, EdgeKind::Normal);
+            //     if let Some(unwind) = unwind {
+            //         self.check_edge(location, *unwind, EdgeKind::Unwind);
+            //     }
+            // }
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_const_eval/src/transform/validate.rs:751]: This one can be safely removed

File: compiler/rustc_const_eval/src/transform/check_consts/resolver.rs [hunk 14]
```diff
@@ -222,17 +222,18 @@ where
...
+        //     let qualif = qualifs::in_operand::<Q, _>(
+        //         self.ccx,
+        //         &mut |l| self.state.qualif.contains(l),
+        //         value,
+        //     );
+
+        //     if !place.is_indirect() {
+        //         self.assign_qualif_direct(place, qualif);
+        //     }
+        // }
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_const_eval/src/transform/check_consts/resolver.rs:236]: Yes, the assignment and drop terminator should work just like the DropAndReplace. We already have tests for

```rust
const fn foo<T>(t: T, mut u: T) -> T {
    u = t;
    u
}
```

and similar cases, so this is even covered by tests.

File: compiler/rustc_middle/src/mir/syntax.rs (hunk 11 above)
Comment (bjorn3) [reply compiler/rustc_middle/src/mir/syntax.rs:585]: > We could even remove the field and check whether the next statement is an assignment to the same place everywhere we need the value of the flag.

That would be fragile to changes in MIR building I think.

File: compiler/rustc_mir_dataflow/src/framework/engine.rs (hunk 4 above)
Comment (Zeegomo) [reply compiler/rustc_mir_dataflow/src/framework/engine.rs]: should be fixed by https://github.com/rust-lang/rust/pull/106430

**Comment by Zeegomo** [comment]
@oli-obk suggestion to use https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html#method.mark_with_reason seems to be very promising. 
I was able to remove `is_replace` entirely and retain all the checks in the borrowck.
There are now a few additional changes in ui tests but most are still regarding the assignment span and I'm confident I can fix the rest without changing everything again.

File: compiler/rustc_mir_build/src/build/scope.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Zeegomo) [reply compiler/rustc_mir_build/src/build/scope.rs:1141]: Followed @oli-obk suggestion and added the terminator later

File: compiler/rustc_mir_dataflow/src/impls/mod.rs [hunk 15]
```diff
@@ -575,9 +590,16 @@ impl<'tcx> GenKillAnalysis<'tcx> for DefinitelyInitializedPlaces<'_, 'tcx> {
     fn terminator_effect(
         &self,
         trans: &mut impl GenKill<Self::Idx>,
-        _terminator: &mir::Terminator<'tcx>,
+        terminator: &mir::Terminator<'tcx>,
         location: Location,
     ) {
+        if let mir::TerminatorKind::Drop { place, .. } = terminator.kind {
```
Comment (arielb1) [inline compiler/rustc_mir_dataflow/src/impls/mod.rs:596]: why this change?

File: compiler/rustc_mir_dataflow/src/impls/mod.rs [hunk 16]
```diff
@@ -458,9 +466,16 @@ impl<'tcx> GenKillAnalysis<'tcx> for MaybeUninitializedPlaces<'_, 'tcx> {
...
```
Comment (arielb1) [inline compiler/rustc_mir_dataflow/src/impls/mod.rs:472]: why this change?

File: compiler/rustc_mir_build/src/build/expr/stmt.rs [hunk 17]
```diff
@@ -39,16 +39,29 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+                    let span = this.tcx.with_stable_hashing_context(|hcx| {
+                        expr_span.mark_with_reason(
+                            None,
+                            DesugaringKind::Replace,
+                            this.tcx.sess.edition(),
+                            hcx,
+                        )
+                    });
+                    let source_info = this.source_info(span);
+                    unpack!(
```
Comment (arielb1) [inline compiler/rustc_mir_build/src/build/expr/stmt.rs:56]: any reason you changed `build_drop_and_replace` to only include the sad path replace?

Comment (Zeegomo) [reply compiler/rustc_mir_build/src/build/expr/stmt.rs:56]: Initially I removed all assigns and added the sad path one later. In the happy path you need the assign anyway at call site because you may not emit a replace if the type does not need a drop, so you can just add the assign regardless of whether you created a new block for a drop and replace or not.

Anyway, not a strong opinion, I'll revert to the original behavior

**Comment by apiraino** [comment]
Switching to waiting on author until all comments are ticked :heavy_check_mark: . Feel free to request a review with `@rustbot ready`, thanks!

@rustbot author

**Comment by Zeegomo** [comment]
Going to close this for now and split into smaller chunks like https://github.com/rust-lang/rust/pull/107271
